bluesky/run_engine.py

```diff
@@ -1353,7 +1353,11 @@ def _open_run(self, msg):
         self._run_start_uids.append(self._run_start_uid)
 
         # Run scan_id calculation method
-        self.md['scan_id'] = self.scan_id_source(self.md)
+        if self.scan_id_source is not None:
+            self.md['scan_id'] = self.scan_id_source(self.md)
```

Comment:
My concern with this is that it would have the side-effect of resetting the `scan_id` to `1` in the HistoryDict, which might not be what the user wants.

Could it be useful provide the `scan_id_source` some way of saying, "Based on the metadata, this scan doesn't get a `scan_id`" without also reseting `scan_id` to `1`? Example sketch to explain what I mean:

Something like

```py
stateful_md = self.md.copy()
if scan_id_source is not None:
    scan_id = scan_id_source(stateful_md)
else:
    scan_id = None
if scan_id is None:  # could be because scan_id_source is not set or because it is set and returned None
    del stateful_md['scan_id']
else:
    self.md['scan_id'] = scan_id
    stateful_md['scan_id'] = scan_id
...
# Combine metadata, in order of decreasing precedence:
md = ChainMap(self._metadata_per_call,  # from kwargs to self.__call__	        md = ChainMap(self._metadata_per_call,  # from kwargs to self.__call__
                      msg.kwargs,  # from 'open_run' Msg	                      msg.kwargs,  # from 'open_run' Msg
                      {'plan_type': plan_type,  # computed from self._plan	                      {'plan_type': plan_type,  # computed from self._plan
                       'plan_name': plan_name},	                       'plan_name': plan_name},
                      stateful_md)  # stateful, persistent metadata	                      self.md)  # stateful, persistent metadata
```

Reply:
Looks great, I'll port this snippet to the PR.